src/driver/m3_subset_test_point_generator.py

```diff
@@ -0,0 +1,110 @@
+    - each of the single publishers,
+    - each of the all-but-one subsets of publishers,
+    - the union set of all publishers,
+    a total of (2p + 1) subsets for training, p being the number of publishers.
+    This class generates the test points that represent the remaining
+    (2^p - 1) - (2p + 1) subsets, for the use case that customers use the
+    planning model to predict the r/f of unobserved subsets.
+
+    When the number of subsets (2^p - 1) - (2p + 1) is too large, randomly
+    sample subsets for testing.
+    """
+
+    def __init__(
+        self,
+        dataset: DataSet,
+        campaign_spend_fractions: np.ndarray,
+        max_num_points: int = 2000,
+        rng: np.random.Generator = np.random.default_rng(0),
+    ):
+        """Constructs an M3SubsetTestPointGenerator.
+
+        Args:
+            dataset:  The DataSet for which test points are to be generated.
+            campaign_spend_fractions:  The campaign_spend / inventory_spend
+                at each publisher, where each inventory_spend is given by
+                dataset._max_spends.
+            max_num_points:  If (2^p - 1) - (2p + 1) > max_num_points, then
+                randomly sample max_num_points subsets.
+            rng:  A numpy Generator object that is used to seed the generation
+                of random test points.
+        """
+        super().__init__(dataset)
+        self._rng = rng
+        self.max_num_points = max_num_points
+        self._campaign_spends = self._max_spends * campaign_spend_fractions
+
+    @staticmethod
+    def subset_layer(p: int, q: int) -> List[np.ndarray]:
+        """Returns a list of subset indicators for a layer of subsets.
+
+        A layer of subsets means the collection of subsets with the same
+        cardinality.
+
+        Args:
+            p: Number of all publishers.
+            q: Number of publishers in the subsets.
+
+        Returns:
+            A list of arrays that iterate over x in {0, 1}^p where sum(x) = q.
+        """
+
+        def one_direction(indices: List) -> np.ndarray:
+            x = np.zeros(p)
+            x[indices] = 1
+            return x
+
+        return [one_direction(list(indices)) for indices in combinations(range(p), q)]
+
+    def test_points(self) -> Iterable[List[float]]:
+        """Returns a generator for generating a list of test points.
+
+        Returns:
+            An iterable of spend vectors representing locations where
+            the true reach surface is to be compared to the modeled reach
+            surface.
+        """
+        p = self._npublishers
+        if p < 4:
+            return
+        if (2 ** p - 1) - (2 * p + 1) > self.max_num_points:
+            n = 0
+            while n < self.max_num_points:
+                subset_indicator = np.random.choice([0, 1], p)
+                num_involved_subsets = np.count_nonzero(subset_indicator)
+                if num_involved_subsets > 1 and num_involved_subsets < p - 1:
```

Comment:
It is equivalent to "1 < num _involved_subsets < p-1".

Reply:
Yep, simplified!